File: Dynamic Programming/1335.Minimum-Difficulty-Of-A-Job-Schedule.py
/*
https://leetcode.com/problems/minimum-difficulty-of-a-job-schedule/
*/

class Solution:
    
    def minDifficulty(self, jobDifficulty: List[int], d: int) -> int:
        
        n = len(jobDifficulty)
        if n < d:
            return -1
        
        jobDifficulty = [-1] + jobDifficulty 
        dp = [[10 ** 9] * (d + 1) for _ in range(n + 1)]
        
        for i in range(1, n + 1):
            for j in range(1, d + 1):
                if j > i:
                    pass
                if j == 1:
                    dp[i][j] = max(jobDifficulty[: i + 1])
                else:
                    for k in range(1, i):
                        dp[i][j] = min(dp[i][j], dp[k][j-1] + max(jobDifficulty[k + 1 : i + 1]))
                        
        return dp[n][d]
    

File: Dynamic Programming/1335.Minimum-Difficulty-Of-A-Job-Schedule.java
/*
https://leetcode.com/problems/minimum-difficulty-of-a-job-schedule/
*/

class Solution {
    public int minDifficulty(int[] jobDifficulty, int d) {
        int n = jobDifficulty.length;
        if(n < d) {
            return -1;
        }
        int[][] dp = new int[n+1][d+1];
        int INF = (int)1e6;
        for(int i = 0; i <= n; i++) {
            for(int j = 0; j <= d; j++) {
                dp[i][j] = INF;
            }
        }
        
        int currMax = 0;
        
        for(int i = 1; i <= n; i++) {
            currMax = Math.max(currMax, jobDifficulty[i-1]);
            dp[i][1] = currMax;
        }
        
        for(int i = 1; i <= n; i++) {
            for(int j = 1; j <= d; j++) {
                for(int k = 1; k < i; k++) {
                    if(j > i) {
                        continue;
                    }
                    dp[i][j] = Math.min(dp[i][j], dp[k][j-1] + getMax(jobDifficulty, k, i-1));
                }
            }
        }    
        return dp[n][d];
    }
    
    
    private int getMax(int[] arr, int start, int end) {
        int ans = 0;
        for(int i = start; i <= end; i++) {
            ans = Math.max(ans, arr[i]);
        }
        return ans;
    }
}
